CUDADataFormats/HcalDigi/interface/DigiCollection.h

```diff
@@ -0,0 +1,187 @@
+
+    static constexpr int WORDS_PER_SAMPLE = 1;
+    static constexpr int HEADER_WORDS = 1;
+
+    static constexpr adc_type adc(uint16_t const* const sample_start) { return (*sample_start & 0xff); }
+
+    static constexpr tdc_type tdc(uint16_t const* const sample_start) { return ((*sample_start >> 8) & 0x3); }
+
+    static constexpr soibit_type soibit(uint16_t const* const sample_start) { return ((*sample_start >> 14) & 0x1); }
+
+    static constexpr uint8_t capid(uint16_t const* const sample_start) { return ((*sample_start >> 10) & 0x3); }
+  };
+
+  struct Flavor4 {
+    static constexpr int WORDS_PER_SAMPLE = 1;
+    static constexpr int HEADER_WORDS = 1;
+  };
+
+  struct Flavor5 {
+    using adc_type = uint8_t;
+
+    static constexpr float WORDS_PER_SAMPLE = 0.5;
+    static constexpr int SAMPLES_PER_WORD = 2;
+    static constexpr int HEADER_WORDS = 1;
+
+    static constexpr adc_type adc(uint16_t const* const sample_start, uint8_t const shifter) {
+      return ((*sample_start >> shifter * 8) & 0x7f);
+    }
+  };
+
+  template <typename Flavor>
+  constexpr uint8_t capid_for_sample(uint16_t const* const dfstart, uint32_t const sample) {
+    auto const capid_first = (*dfstart >> 8) & 0x3;
+    return (capid_first + sample) & 0x3;  // same as % 4
+  }
+
+  template <>
+  constexpr uint8_t capid_for_sample<Flavor3>(uint16_t const* const dfstart, uint32_t const sample) {
+    return Flavor3::capid(dfstart + Flavor3::HEADER_WORDS + sample * Flavor3::WORDS_PER_SAMPLE);
+  }
+
+  template <typename Flavor>
+  constexpr typename Flavor::soibit_type soibit_for_sample(uint16_t const* const dfstart, uint32_t const sample) {
+    return Flavor::soibit(dfstart + Flavor::HEADER_WORDS + sample * Flavor::WORDS_PER_SAMPLE);
+  }
+
+  template <typename Flavor>
+  constexpr typename Flavor::adc_type adc_for_sample(uint16_t const* const dfstart, uint32_t const sample) {
+    return Flavor::adc(dfstart + Flavor::HEADER_WORDS + sample * Flavor::WORDS_PER_SAMPLE);
+  }
+
+  template <typename Flavor>
+  constexpr typename Flavor::tdc_type tdc_for_sample(uint16_t const* const dfstart, uint32_t const sample) {
+    return Flavor::tdc(dfstart + Flavor::HEADER_WORDS + sample * Flavor::WORDS_PER_SAMPLE);
+  }
+
+  template <>
+  constexpr Flavor5::adc_type adc_for_sample<Flavor5>(uint16_t const* const dfstart, uint32_t const sample) {
+    // avoid using WORDS_PER_SAMPLE and simply shift
+    return Flavor5::adc(dfstart + Flavor5::HEADER_WORDS + (sample >> 1), sample % 2);
+  }
+
+  template <typename Flavor>
+  constexpr uint32_t compute_stride(uint32_t const nsamples) {
+    return static_cast<uint32_t>(nsamples * Flavor::WORDS_PER_SAMPLE) + Flavor::HEADER_WORDS;
+  }
+
+  template <typename Flavor>
+  constexpr uint32_t compute_nsamples(uint32_t const nwords) {
+    return (nwords - Flavor::HEADER_WORDS) / Flavor::WORDS_PER_SAMPLE;
+  }
+
+  template <>
+  constexpr uint32_t compute_nsamples<Flavor5>(uint32_t const nwords) {
```

Comment:
I would say this is to avoid a division of an integer by 0.5, which would involve floating point operations - while multiplying by 2 can be done with integer operations.

Reply:
What I meant is: why cannot you use the previous templated `compute_nsamples` also for `Flavor = Flavor5` ? (Then I agree with you that in all cases one could simply multiply by `SAMPLES_PER_WORD` instead of dividing by `WORDS_PER_SAMPLE`)

Reply:
the other flavors were removed for which WORDS_PER_SAMPLE was 2 (flavor 2), which means you again have floating point in there... i just tried to keep only int operations in there...

Reply:
@perrotta
> why cannot you use the previous templated `compute_nsamples` also for `Flavor = Flavor5` ?

The generic version is (lines 89-92)
```c++
  template <typename Flavor>
  constexpr uint32_t compute_nsamples(uint32_t const nwords) {
    return (nwords - Flavor::HEADER_WORDS) / Flavor::WORDS_PER_SAMPLE;
  }
```

For `Flavor5` that would become
```c++
  constexpr uint32_t compute_nsamples(uint32_t const nwords) {
    return (nwords - 1) / 0.5;
  }
```
which does not look very sensible.

The `Flavor5` specialisation instead becomes
```c++
  constexpr uint32_t compute_nsamples(uint32_t const nwords) {
    return (nwords - 1) * 2;
  }
```
which avoids floating point math for an integer multiplication by 2.


> Then I agree with you that in all cases one could simply multiply by `SAMPLES_PER_WORD` instead of dividing by `WORDS_PER_SAMPLE`

I have no idea if - in principle - there can be cases where `WORDS_PER_SAMPLE` is greater than 1.
If that is not foreseen, then indeed we can change the generic version to
```c++
  template <typename Flavor>
  constexpr uint32_t compute_nsamples(uint32_t const nwords) {
    return (nwords - Flavor::HEADER_WORDS) * Flavor::SAMPLES_PER_WORD;
  }
```

If we want to keep this more generic, it could become
```c++
  template <typename Flavor>
  constexpr uint32_t compute_nsamples(uint32_t const nwords) {
    if constexpr(Flavor::SAMPLES_PER_WORD >= 1)
      return (nwords - Flavor::HEADER_WORDS) * Flavor::SAMPLES_PER_WORD;
    else
      return (nwords - Flavor::HEADER_WORDS) / Flavor::WORDS_PER_SAMPLE;
  }
```